### ISSUE TITLE
SQL Statements for Deleting a Principle/Parent in an Optional Relationship when the OnDelete Behavior is SetNull

### ISSUE THREAD

**Issue by jaliyaudagedara** [dotnet/efcore#28503]
**EF Core Version**: **7.0.0-preview.6.22329.4**
**Database Provider**: **Microsoft SQL Server**

Consider the following simple example.
```C#
public class Category
{
    public int Id { get; set; }
    public string Name { get; set; }

    public List<ToDo> ToDos { get; set; }
}

public class ToDo
{
    public int Id { get; set; }
    public string Title { get; set; }

    public int? CategoryId { get; set; }
    public Category Category { get; set; }
}
```
`Principle/Parent` is configured as follows.
```C#
modelBuilder.Entity<Category>()
    .HasMany(x => x.ToDos)
    .WithOne(x => x.Category)
    .HasForeignKey(x => x.CategoryId)
    .OnDelete(DeleteBehavior.SetNull);
```
And say I am deleting a `Category` and for some reason, its `ToDos` is loaded into the `DbContext`.
```C#
Category category = await context.Categories
    .Include(x => x.ToDos)
    .FirstOrDefaultAsync();

context.Categories.Remove(category);
await context.SaveChangesAsync();
```
In this case, the generated SQL looks like below.
```
--Executed DbCommand (4ms) [Parameters=[@p1='1', @p0=NULL (DbType = Int32), @p2='1'], CommandType='Text', CommandTimeout='30']
SET NOCOUNT ON;

UPDATE [ToDos] SET [CategoryId] = @p0
OUTPUT 1
WHERE [Id] = @p1;

DELETE FROM [Categories]
OUTPUT 1
WHERE [Id] = @p2;
```

**My question is**, in this case, do we need the `UPDATE` statement? Since the `OnDelete` behavior is `DeleteBehavior.SetNull`, the FK is created with `ON DELETE SET NULL` and the database is aware of what to do with `Dependents/Children`, so only the `DELETE` statement is enough.
```
CREATE TABLE [ToDos] (
    [Id] int NOT NULL IDENTITY,
    [Title] nvarchar(max) NULL,
    [CategoryId] int NULL,
    CONSTRAINT [PK_ToDos] PRIMARY KEY ([Id]),
    CONSTRAINT [FK_ToDos_Categories_CategoryId] FOREIGN KEY ([CategoryId]) REFERENCES [Categories] ([Id]) ON DELETE SET NULL
);
```

**Comment by ajcvickers** [comment]
@jaliyaudagedara EF always performs explicit updates for any entity that is tracked to ensure that the database and the tracked entities remain in sync, even when the database does not support or has not been configured with cascading options, since cascades in the database are generally much more limited than those supported by EF.

**Comment by jaliyaudagedara** [comment]
Thanks @ajcvickers for the explanation.

When we have not loaded `Dependents/Children`, I saw EF was using only the `DELETE` statement and since we have the `ON DELETE SET NULL` in the constraint, it's updating the FK in `Dependents/Children` to null. I was thinking even with `Dependents/Children` are loaded, we could do the same (only when OnDelete behavior is `DeleteBehavior.SetNull`).

It's not a big deal though, there are no multiple roundtrips, just an additional DELETE statement in the same trip.

Closing the issue.

Thanks again.
